randomizers/starting_item.py

```diff
@@ -0,0 +1,147 @@
+    available_items: set[str] = set()
+    for item, fraction in items_by_usefulness.items():
+      group_name = next(
+        (group for group, group_items in self.logic.progress_item_groups.items() 
+         if item in group_items),
+        None
+      )
+      # However we only want to merge the group if all the items are progression
+      # (specifically for pearls where Farore's can be useful without the
+      # others being progression)
+      if (
+        group_name and 
+        all((item in self.logic.get_flattened_unplaced_progression_items()) for item in self.logic.progress_item_groups[group_name])
+      ):
+        item = group_name
+        fraction = min((items_by_usefulness[item]) for item in self.logic.progress_item_groups[group_name])
+        fraction -= (len(self.logic.progress_item_groups[group_name]) - 1)
+
+      if fraction <= max_fraction:
+        available_items.add(item)
+
+    available_items -= DISALLOWED_RANDOM_STARTING_ITEMS
+    if not "Delivery Bag" in self.logic.currently_owned_items:
+      # Delivery bag is the only bag that can hold progression items, and we
+      # don't want to give a progression item if we don't have its bag since
+      # it's impossible to see the item until you get delivery bag in that case
+      available_items -= DELIVERY_BAG_ITEMS
+    if not self.options.get("keylunacy"):
+      available_items -= set(DUNGEON_PROGRESS_ITEMS)
+
+    # To avoid treasure charts overwhelming everything when enabled, group them so they have the same weight as any other item
+    if set(self.logic.treasure_chart_names).intersection(available_items):
+        available_items -= set(self.logic.treasure_chart_names)
+        available_items.add('Treasure Chart')
+    if set(self.logic.triforce_chart_names).intersection(available_items):
+        available_items -= set(self.logic.triforce_chart_names)
+        available_items.add('Triforce Chart')
+
+    # If sword_mode is No Starting Sword, adding swords would desync the
+    # sword_mode option from the actual number of swords.
+    # Swordless doesn't put the Progressive Swords in the progress items in the
+    # first place so doesn't need this
+    # This might be removable if we can make a random starting progressive sword
+    # automatically change the sword_mode variable and its consequences
+    if self.options.get("sword_mode") == "No Starting Sword" and "Progressive Sword" in available_items:
+      available_items.remove("Progressive Sword")
+
+    # without shuffling here, the chosen items sometimes stay stable for the
+    # same seed with different starting_gear options as a prefix of the list may
+    # be the same.
+    # We need to sort before converting to lists because sets have per-process
+    # hash seeding and could otherwise lead to different randomization between
+    # processes with the same seed
+    return self.rng.sample(sorted(available_items), k=len(available_items))
```

Comment:
Is this call the rng.sample necessary? You can make a set deterministic just by calling sorted, not sure what is accomplished by sampling the entire list and then randomly choosing from it, and the comment doesn't explain it either.

Reply:
It's not necessary for correctness, I can drop it if undesirable

Without it, the only call(s) to the rng is the `self.rng.choice` to select the actual item, and the list that gets passed to it is fairly stable (there isn't a huge amount of variation in progression items, at least with the settings I've tried), so with the same seed but changing settings, you would still very often get the exact same set of starting items. The `sample` makes it so adding or removing a single item from the `available_items` list results in completely different selected items even with the same seed

I don't know how much you care about things changing with the same seed when changing settings, so this is possibly just not relevant and can be dropped

Reply:
I actually usually try to aim for the opposite of what you described, so that changing settings while keeping the seed the same changes the randomization as little as possible. For example, I explicitly reset the RNG seed in between each separate randomizer option being run to minimize how much they affect each other and make it mostly depend on the seed (which is probably part of the reason you saw these starting items being the same with different settings). Though the options affect each other in so many indirect ways that it's not really possible to make it depend entirely on the seed in most cases, so this isn't an explicit feature.

Also, for cases where making a list more random actually is desired, I would just call rng.shuffle on that list and then pop one from it, which is probably more readable and performant than sampling the entire list from itself and then choosing one from it.

Reply:
Understood! I'll drop it and update the PR